devices/usbc/ctrl.c

```diff
@@ -0,0 +1,410 @@
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "../low.h"
+#include "../errors.h"
+
+#include "client.h"
+#include "usbphy.h"
+
+
+struct {
+	u32 qtdOffs;
+
+	usb_dc_t *dc;
+	usb_common_data_t *data;
+} ctrl_common;
+
+
+/* device cotroller register offsets */
+enum {
+	/* identification regs */
+	id = 0x0, hwgeneral, hwhost, hwdevice, hwtxbuf, hwrxbuf,
+
+	/* operational regs */
+	gptimer0ld  = 0x20, gptimer0ctrl, gptimer1ld, gptimer1ctrl, sbuscfg,
+
+	/* capability regs */
+	caplength = 0x40, hciversion = 0x40, hcsparams, hccparams,
+	dciversion = 0x48, dccparams,
+
+	/* operational regs cont. */
+	usbcmd = 0x50, usbsts, usbintr, frindex,
+	periodiclistbase = 0x55, deviceaddr = 0x55, asynclistaddr = 0x56,
+	endpointlistaddr = 0x56, burstsize = 0x58, txfilltunning, endptnak = 0x5E,
+	endptnaken, configflag, portsc1, otgsc = 0x69, usbmode, endptsetupstat,
+	endptprime, endptflush, endptstat, endptcomplete, endptctrl0, endptctrl1,
+	endptctrl2, endptctrl3, endptctrl4, endptctrl5, endptctrl6, endptctrl7
+};
+
+
+static int ctrl_allocBuff(int endpt, int dir)
+{
+	ctrl_common.data->endpts[endpt].buf[dir].buffer = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+	if (ctrl_common.data->endpts[endpt].buf[dir].buffer == NULL)
+		return ERR_MEM;
+
+	ctrl_common.data->endpts[endpt].buf[dir].len = 0;
+
+	return ERR_NONE;
+}
+
+
+static void *ctrl_allocQtdMem(void)
+{
+	int i;
+
+	if (!ctrl_common.qtdOffs) {
+		ctrl_common.dc->dtdMem = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+		if (ctrl_common.dc->dtdMem == NULL)
+			return NULL;
+
+		for (i = 0; i < USB_BUFFER_SIZE; ++i)
+			((char *)ctrl_common.dc->dtdMem)[i] = 0;
+	}
+
+	return  (void *)(ctrl_common.dc->dtdMem + 0x40 * (ctrl_common.qtdOffs++));
+}
+
+
+static int ctrl_dtdInit(int endpt, int inQH, int outQh)
+{
+	dtd_t *dtd;
+	int qh = endpt * 2;
+
+	if (!endpt)
+		return -1;
+
+	/* initialize dtd for selected endpoints */
+	if (outQh) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[qh].base = (((u32)dtd) & ~0xfff);
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd;
+		ctrl_common.dc->endptqh[qh].tail = dtd;
+	}
+
+	if (inQH) {
+		dtd = ctrl_allocQtdMem();
+		if (dtd == NULL)
+			return ERR_MEM;
+
+		ctrl_common.dc->endptqh[++qh].base = (((u32)dtd) & ~0xfff) + (64 * sizeof(dtd_t));
+		ctrl_common.dc->endptqh[qh].size = 0x40;
+		ctrl_common.dc->endptqh[qh].head = dtd + 64;
+		ctrl_common.dc->endptqh[qh].tail = dtd + 64;
+	}
+
+	return ERR_NONE;
+}
+
+
+static int ctrl_initEndptQh(int endpt, int dir, endpt_data_t *endpt_init)
+{
+	u32 setup = 0;
+	int qh = endpt * 2 + dir;
+
+	ctrl_allocBuff(endpt, dir);
+
+	ctrl_common.dc->endptqh[qh].caps =  endpt_init->caps[dir].max_pkt_len << 16;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].ios << 15;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].zlt << 29;
+	ctrl_common.dc->endptqh[qh].caps |= endpt_init->caps[dir].mult << 30;
```

Comment:
```suggestion
	u32 tmp;
	tmp = endpt_init->caps[dir].max_pkt_len << 16;
	tmp |= endpt_init->caps[dir].ios << 15;
	tmp |= endpt_init->caps[dir].zlt << 29;
	tmp |= endpt_init->caps[dir].mult << 30;
	ctrl_common.dc->endptqh[qh].caps = tmp;
```
What do you think about this solution.
It will be simpler (smaller size ) and faster.
`endptqh` is a pointer to `dqh_t` structure, which is with attribute `packed`.
From the GCC documentation:
```
-munaligned-access
-mno-unaligned-access

    Enables (or disables) reading and writing of 16- and 32- bit values from addresses that are not 16- or 32- bit aligned. By default unaligned access is disabled for all pre-ARMv6, all ARMv6-M and for ARMv8-M Baseline architectures, and enabled for all other architectures. If unaligned access is not enabled then words in packed data structures are accessed a byte at a time.

    The ARM attribute Tag_CPU_unaligned_access is set in the generated object file to either true or false, depending upon the setting of this option. If unaligned access is enabled then the preprocessor symbol __ARM_FEATURE_UNALIGNED is also defined.
```

Reply:
Done